File: services/ai-service/src/services/insightsService.js
const { callDeepSeek } = require('./deepseekService');
const { buildSystemContext } = require('./contextBuilder');

const CACHE_TTL = 5 * 60 * 1000; // 5 minutos
let cache = null;
let cacheExpiry = 0;

/**
 * Genera insights automáticos del negocio con caché.
 * @returns {Promise<{insight:string,trend_percentage:number,trend_direction:string,trend_comparison:string,timestamp:string}>}
 */
async function generateInsights() {
  const now = Date.now();
  if (cache && now < cacheExpiry) return cache;

  const contextData = await buildSystemContext().catch(() => null);

  const systemPrompt = `Eres un analista de negocio para Kiora POS. Genera exactamente 1 insight accionable basado en los datos actuales.

REGLAS:
- El insight debe ser breve, concreto y accionable (máximo 200 caracteres)
- Calcula la tendencia de rendimiento comparando con la semana pasada
- El insight debe mencionar datos específicos del contexto, no genéricos
- Responde ÚNICAMENTE con un objeto JSON válido, sin markdown ni texto adicional

Formato de respuesta:
{
  "insight": "texto del insight aquí, con datos específicos",
  "trend_percentage": 18,
  "trend_direction": "up",
  "trend_comparison": "vs semana pasada"
}

Datos actuales del negocio:
${contextData || 'No hay datos disponibles en este momento.'}`;

  try {
    const result = await callDeepSeek(
      [{ role: 'system', content: systemPrompt }],
      [] // sin tools para respuestas más rápidas
    );

    let parsed;
    try {
      parsed = JSON.parse(result.content);
    } catch {
      // Fallback: extraer JSON del texto
      const match = result.content.match(/\{[\s\S]*\}/);
      if (match) parsed = JSON.parse(match[0]);
      else throw new Error('No se pudo parsear la respuesta');
    }

    const output = {
      insight: parsed.insight || 'No hay suficientes datos para generar un insight en este momento.',
      trend_percentage: parsed.trend_percentage ?? 0,
      trend_direction: parsed.trend_direction === 'down' ? 'down' : 'up',
      trend_comparison: parsed.trend_comparison || 'vs semana pasada',
      timestamp: new Date().toISOString(),
    };

    cache = output;
    cacheExpiry = now + CACHE_TTL;
    return output;
  } catch (err) {
    // Si hay cache expirado pero tenemos datos, servirlos como fallback
    if (cache) return cache;
    throw err;
  }
}

/** Limpia el caché forzadamente */
function clearCache() {
  cache = null;
  cacheExpiry = 0;
}

module.exports = { generateInsights, clearCache };

File: services/ai-service/app/services/insights_service.py
import time
import json
import logging
from datetime import datetime
from app.services.deepseek_service import client, DEEPSEEK_MODEL
from app.services.context_builder import build_system_context
from app.models.schemas import InsightResponse

logger = logging.getLogger(__name__)

CACHE_TTL = 300  # 5 minutos
_cache = None
_cache_expiry = 0

async def generate_insights() -> dict:
    global _cache, _cache_expiry
    now = time.time()
    
    if _cache and now < _cache_expiry:
        return _cache

    try:
        context_data = await build_system_context()
    except Exception:
        context_data = None

    system_prompt = f"""Eres un analista de negocio para Kiora POS. Genera exactamente 1 insight accionable basado en los datos actuales.

REGLAS:
- El insight debe ser breve, concreto y accionable (máximo 200 caracteres)
- Calcula la tendencia de rendimiento comparando con la semana pasada
- El insight debe mencionar datos específicos del contexto, no genéricos
- Responde ÚNICAMENTE con un objeto JSON válido, sin markdown ni texto adicional
- El objeto JSON DEBE tener exactamente la siguiente estructura (y no otras claves):
{{
  "insight": "El texto del insight...",
  "trend_percentage": 15.5,
  "trend_direction": "up",
  "trend_comparison": "vs semana pasada"
}}
Nota: trend_direction debe ser "up" o "down". trend_percentage debe ser un número.

Datos actuales del negocio:
{context_data or 'No hay datos disponibles en este momento.'}"""

    try:
        response = await client.chat.completions.create(
            model=DEEPSEEK_MODEL,
            messages=[{"role": "system", "content": system_prompt}],
            max_tokens=4096,
            stream=False,
            response_format={"type": "json_object"}
        )
        
        content = response.choices[0].message.content or "{}"
        
        # Validación estricta con Pydantic
        validated_data = InsightResponse.model_validate_json(content)
        
        output = validated_data.model_dump()
        output["timestamp"] = datetime.utcnow().isoformat() + "Z"

        _cache = output
        _cache_expiry = now + CACHE_TTL
        return output
    except Exception as e:
        logger.error(f"[InsightsService] Error: {e}")
        if _cache:
            return _cache
        raise e

def clear_cache():
    global _cache, _cache_expiry
    _cache = None
    _cache_expiry = 0
